File: server/amrit_central_server/central/retention.py
"""Retention: expiry of row-level operational data the server holds.

The server is aggregate-first — row-level isolate data never leaves the site — but it does
accumulate an operational trail: the queries it issued, what came back, and the poll audit.
Several jurisdictions cap how long that may be kept, and until now nothing expired.

Deliberately NOT purged:

* ``KPISnapshot`` — the aggregate surveillance record itself. It is what the system exists
  to produce, carries no row-level personal data, and deleting it would silently rewrite
  published history.
* ``Site``, ``UserProfile``, geography, roles and configuration. These describe the
  deployment, not a person, and removing them would break referential integrity.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import timedelta

from django.db import transaction
from django.utils import timezone

from .privacy import retention_days


@dataclass(frozen=True)
class RetentionTarget:
    label: str
    model_path: str
    #: Field carrying the row's age. Always the latest available, never the earliest.
    field: str


#: ``QueryDispatch`` is deliberately absent: it carries no timestamp of its own that is
#: always set, and it cascades from its ``Query``, so it expires with the query it belongs
#: to rather than on a date it cannot supply.
RETENTION_TARGETS: tuple[RetentionTarget, ...] = (
    RetentionTarget("Poll audit entries", "queries.PollAuditEntry", "created_at"),
    RetentionTarget("Query results", "queries.QueryResult", "received_at"),
    RetentionTarget("Queries", "queries.Query", "created_at"),
    RetentionTarget("Dashboard refresh runs", "dashboards.DashboardRefreshRun", "clicked_at"),
)


def cutoff_for(days: int | None, *, now=None):
    """The instant on or after which data is kept, or None when retention is unset."""
    if days is None:
        return None
    if int(days) < 1:
        raise ValueError(f"Retention must be at least 1 day: {days!r}")
    return (now or timezone.now()) - timedelta(days=int(days))


def _model(path: str):
    from django.apps import apps

    app_label, model_name = path.split(".", 1)
    return apps.get_model(app_label, model_name)


def purge_expired(*, country_code: str | None = None, days: int | None = None, dry_run: bool = True, now=None) -> dict:
    """Report — and when ``dry_run`` is false, delete — data past the retention period.

    Ordered child-before-parent so a cascade never removes a row this pass has already
    counted, and wrapped in one transaction so a failure leaves nothing half-purged.
    """
    period = days if days is not None else retention_days(country_code)
    cutoff = cutoff_for(period, now=now)
    if cutoff is None:
        return {"applied": False, "dry_run": dry_run, "retention_days": None, "cutoff": None, "removed": []}

    removed = []

    def run() -> None:
        for target in RETENTION_TARGETS:
            queryset = _model(target.model_path).objects.filter(**{f"{target.field}__lt": cutoff})
            count = queryset.count()
            removed.append({"label": target.label, "model": target.model_path, "rows": count})
            if not dry_run and count:
                queryset.delete()

    if dry_run:
        run()
    else:
        with transaction.atomic():
            run()

    return {
        "applied": True,
        "dry_run": dry_run,
        "retention_days": int(period),
        "cutoff": cutoff.isoformat(),
        "removed": removed,
    }

File: app/src/core/domain/retention.ts
/**
 * Retention policy: profile-driven expiry of row-level operational data.
 *
 * Several jurisdictions cap how long identifiable surveillance data may be kept, and today
 * nothing in this application expires. `profile.privacy.retention_days` sets that cap.
 * Leaving it null retains indefinitely, which is exactly what every existing deployment
 * does now — so adding retention changes nothing until a deployment asks for it.
 *
 * Deliberately NOT purged:
 *
 *  - The One Health audit chain. Deleting entries would break the hash chain and destroy
 *    the evidence that the log is intact. Personal data inside an entry is removed by
 *    `eraseAuditDetails()` instead, which shreds the payload and leaves the chain
 *    verifiable. Retention and tamper-evidence are reconciled there, not here.
 *  - `national_outbox`. It carries aggregates, which are not row-level personal data, and
 *    dropping a queued item would lose a report the central server never received.
 *  - Catalogue, master and configuration data. These describe the deployment, not a person.
 */

export interface RetentionDependent {
  table: string
  /** Column in the dependent table holding the parent row's id. */
  column: string
}

export interface RetentionTarget {
  table: string
  /** Primary-key column, used to select the rows to remove. */
  key: string
  label: string
  /**
   * SQL yielding the row's retention anchor as a `YYYY-MM-DD` string.
   *
   * Always the *latest* of the candidate dates, never the earliest. A record edited last
   * week is in use whatever its specimen date says, and a mistyped or non-ISO specimen date
   * ('01/02/2020', a 1925 typo) sorts below every plausible cutoff — taking the maximum
   * means such a row is held by its `created_at` rather than silently destroyed.
   *
   * Compared on the date prefix alone because these columns mix `CURRENT_TIMESTAMP`
   * ('2026-08-12 09:30:00') with ISO instants ('2026-08-12T09:30:00.000Z'), which do not
   * order against each other below day granularity. Retention in days needs no finer.
   */
  anchor: string
  /** Rows that reference this one and must be removed with it (no ON DELETE CASCADE). */
  dependents?: RetentionDependent[]
}

const dateOf = (column: string): string => `substr(coalesce(${column},''),1,10)`

export const RETENTION_TARGETS: readonly RetentionTarget[] = Object.freeze([
  {
    table: 'isolates',
    key: 'id',
    label: 'Isolate records',
    // AST results, genomic results and omics rows carry ON DELETE CASCADE and go with the
    // isolate; foreign keys are enabled on every connection (see initialize()).
    anchor: `max(${dateOf('created_at')},${dateOf('updated_at')},${dateOf('specimen_date')})`
  },
  {
    table: 'national_events',
    key: 'id',
    label: 'One Health events',
    anchor: `max(${dateOf('recorded_at')},${dateOf('observed_at')})`,
    // Both reference national_events without a cascade, so they are removed explicitly.
    // An alert message and an action's evidence quote the event, so they expire with it.
    dependents: [
      { table: 'national_alerts', column: 'event_id' },
      { table: 'national_actions', column: 'event_id' }
    ]
  }
])

export class RetentionError extends Error {}

/**
 * The date on or after which data is kept. Rows whose anchor is strictly earlier expire.
 *
 * Returns null when retention is unset, which means "keep indefinitely" rather than
 * "expire everything" — the failure mode of the opposite default is unrecoverable.
 */
export function retentionCutoffDate(
  retentionDays: number | null | undefined,
  now: Date = new Date()
): string | null {
  if (retentionDays === null || retentionDays === undefined) return null
  if (!Number.isFinite(retentionDays) || !Number.isInteger(retentionDays) || retentionDays < 1) {
    throw new RetentionError(`Retention must be a whole number of days of at least 1: ${String(retentionDays)}`)
  }
  const cutoff = new Date(now.getTime() - retentionDays * 24 * 60 * 60 * 1000)
  if (Number.isNaN(cutoff.getTime())) throw new RetentionError('Retention cutoff is not a valid date')
  return cutoff.toISOString().slice(0, 10)
}

/** Selects the expired rows of one target. Bound parameter: the cutoff date. */
export function expiredRowsSql(target: RetentionTarget): string {
  // An empty anchor means the row carries no usable date at all. It is kept: retention
  // removes data proven to be old, and an undated row proves nothing.
  return `SELECT ${target.key} AS id FROM ${target.table} WHERE ${target.anchor} <> '' AND ${target.anchor} < ?`
}
